### ISSUE TITLE
[question] Should a test_package be able to change options of a dependency of the recipe under test?

### ISSUE THREAD

**Issue by madebr** [conan-io/conan#7547]
A test package can change options on a dependency, that is also used by the recipe.
This changes the behavior of running conan with tests enabled or disabled.

Consider the following `rec1` and `rec2` recipe and test of rec2.
```sh
cat >rec1.py <<EOF
from conans import ConanFile
class Rec1Conan(ConanFile):
    name = "rec1"
    version = "1.0"
    options = {
        "opt": [True, False],
    }
    default_options = {
        "opt": True,
    }
EOF

cat >rec2.py <<EOF
from conans import ConanFile
class Rec2Conan(ConanFile):
    name = "rec2"
    version = "1.0"

    requires = "rec1/1.0"
EOF

mkdir tp_rec2

cat >tp_rec2/conanfile.py <<EOF
from conans import ConanFile
class TestRec2Conan(ConanFile):
    def configure(self):
        self.options["rec1"].opt = False
EOF
```

Run and compare following series of commands:
```sh
conan remove -f rec1
conan remove -f rec2
conan create rec1.py rec1/1.0@ -tf None
conan create rec2.py rec2/1.0@ -tf None
```
versus:
```sh
conan remove -f rec1
conan remove -f rec2
conan create rec1.py rec1/1.0@ -tf None
conan create rec2.py rec2/1.0@ -tf tp_rec2
```

The second set of commands failed because the test package changed an option of a dependency that modified its package id.

I think this is undefined behavior and conan should error when this is detected.

**Comment by jgsogo** [comment]
I agree this is totally undesired, I cannot imagine a scenario where this would be needed. `test_package` should use the packages that has been built or fail, I mean, it should consume the same lockfile. 

Your example is the good one because it raises an error, but, what if you are overriding a dependency? You have compiled your package using a different version 😬 

OTH the responsibility is on the recipe-author side, it is not something affecting consumers.

Thanks for the issue!

**Comment by jmarrec** [comment]
for transparency, this came up while @madebr was helping me with a build error of mine and we looked at why a certain dependency (which fails to buil on x86) was pulled early on, when really it's only needed in the requirements of my test_package.

https://github.com/NREL/conan-openstudio-ruby/blob/57460a1fff0ebc54d2bbbe61b8979c3475c8bd01/test_package/conanfile.py#L20-L26

Ideally I'd move everything as a build_requirements there, but 1) swig as build_requirements somehow makes my build fail 2) You can't use (it seems) something like `build_requires_option['zlib'].minizip=True`

**Comment by memsharded** [comment]
This was changed in Conan 2.0, the ``test_pacakge`` is a separate and independent thing, so it cannot change anything in the package being created (and tested). The result must be identical using or not the ``test_package`` (``-tf=""`` to disable it). Closing as solved in 2.0
